### ISSUE TITLE
Add support for notification

### ISSUE THREAD

**Issue by marc0x1** [GeopJr/Tooth#42]
For both desktop and mobile Linux, https://unifiedpush.org is a standard that can deliver push notifications over D-Bus, while allowing the end-user to self-host infrastructure. UnifiedPush on Android is already being used by a few apps like  Tusky and Fedilab ( mastodon clients ) , but the Linux implementation is new.
The UP chat is at https://matrix.to/#/#unifiedpush:matrix.org

It's probably a low-priory feature, but it's a nice thing to have.

**Comment by johnlpe** [comment]
_From #349:_ 

Please add support for [UnifiedPush](https://unifiedpush.org) push notification protocol/standard.

We need push notifications, but we don’t want use centralized online service that read we all notifications, like Android has Google Firebase Messaging Service and iOS has Apple Push Notification service. UnifiedPush protcol is the solution for this. UnifiedPush is open protocol/standard for connection between online service (Application Server) → notification service (Push Server) and notification service’s client (Push Distributor) → application (End User Application). It enables the user to choose the service through he receives his push notifications.

Especially GNU/Linux on mobile is very big need of push notifications and with UnifiedPush we can fix it.

<img src="https://raw.githubusercontent.com/UnifiedPush/documentation/main/static/img/animation-flow.svg" width="300"/> <img style="float: right;" src="https://github.com/mastodon/mastodon/assets/138632900/a52dd1a5-c972-4e00-8e65-4fec1aa3a5a0" width="300" />

**Comment by johnlpe** [comment]
> I don't understand why you can keep Tuba running in the background. If this feature isn't supported, what's the point of that?

@GeopJr

**Comment by johnlpe** [comment]
Tuba does support notifications when the app is the foreground, right? And remember the point of UnifiedPush is that one Push Distributor keep active connection. When Push Distributor receive notification it shows notifications directly or wakes up End User Application like Tuba (the notification can be only `ping` that isn't shown to the user and the End User Application fetch the content of the notification from the server after being woken up by UnifiedPush).

**Comment by GeopJr** [comment]
> I don't understand why you can keep Tuba running in the background. If this feature isn't supported, what's the point of that?

The renaming of this issue should make it a bit more clear:

- Tuba has notifications, as long as it's running (either on the foreground or the background (with `--hidden` or when "background work" is enabled etc))
- UnifiedPush allows you to receive notifications without Tuba constantly running / keeping an active websocket connection to your instance

This issue is low priority however.

**Comment by johnlpe** [comment]
> UnifiedPush allows you to receive notifications without Tuba constantly running / keeping an active websocket connection to your instance

Yup, and I think supporting UnifiedPush it pretty trivial. The standard defines a quite simple way to receive notifications and I think UnifiedPush contributors on https://matrix.to/#/#unifiedpush:matrix.org are happy to help. Also tehere is KUnifiedPush library by KDE.

https://unifiedpush.org/spec/dbus/
https://invent.kde.org/libraries/kunifiedpush
https://www.volkerkrause.eu/2022/11/12/kde-unifiedpush-push-notifications.html

> This issue is low priority however.

Why you think so? I think this is very important for GNU/Linux on mobile.

**Comment by GeopJr** [comment]
> Why you think so? I think this is very important for GNU/Linux on mobile.

Someone (probably me) needs to write bindings to the C/Go library and the wording on all related pages doesn't sound that promising:

> NOTE: This library is unstable as long as it's on version 0.x.x
(https://github.com/UnifiedPush/go_dbus_connector)

> Linux support is still experimental, also check out [KUnifiedPush](https://invent.kde.org/libraries/kunifiedpush/) for an application-framework (KDE) based implementation.
(https://unifiedpush.org/developers/go_c/)

**Comment by johnlpe** [comment]
> Someone (probably me) needs to write bindings to the C/Go library and the wording on all related pages doesn't sound that promising:

Eh, the killer feature for v2.0 of Tuba?:laughing:

**Comment by GeopJr** [comment]
Could you open a new bug issue for it? Preferably with logs (`flatpak run dev.geopjr.Tuba`) - specifically any errors before or after messages that start with "Opening stream: ". FWIW, notifications have been supported since Tootle / before Tuba and they use the portal so something else is going on.

**Comment by GeopJr** [comment]
> Do you think a libunifiedpush (.so) would help its adoption? Would it help you to support it? Or, as the protocol is pretty simple, do you think it would be more efficient to get an implementation per framework directly?

Hey! On GNOME-land, I believe it would, but specifically with gobject-introspection (so all languages in the ecosystem (python, vala, js, rust etc) get automatically generated bindings). Now that the dbus doc feels more complete than the last time I gave it a try, even if you aim for a non-glib-based library, someone could easily create one with some additional helper functions.

> Would it help you to support it?

I've been pretty busy since the last release so I can't make any promises. Since this issue was opened, Tuba's notifications have changed a bit, from filtering to multi-account aware notification quick actions. Judging from Android apps, this is something UnifiedPush doesn't prevent in anyway but it requires more studying from my side which I don't have the time for at the moment.

**Comment by p1gp1g** [comment]
For information, I've just published [*libunifiedpush*](https://codeberg.org/UnifiedPush/unifiedpush-rs/src/branch/main/libunifiedpush), a new library with GObject-introspection for UnifiedPush

There might be a few things to improve, we'll see when the first apps will start using it. I've tested from C and python
